File: internal/graphrag/queries.go
package graphrag

import (
	"context"
	"math"
	"sort"
	"time"
)

// ErrorChain traces error spans upstream to find the root cause service.
// The tenant slice is selected via ctx — callers without a tenant ctx
// collapse to storage.DefaultTenantID at the coordinator boundary.
func (g *GraphRAG) ErrorChain(ctx context.Context, service string, since time.Time, limit int) []ErrorChainResult {
	if limit <= 0 {
		limit = 10
	}

	stores := g.storesFor(ctx)

	errorSpans := stores.traces.ErrorSpans(service, since)
	if len(errorSpans) > limit {
		errorSpans = errorSpans[:limit]
	}

	var results []ErrorChainResult
	seen := make(map[string]bool)

	for _, span := range errorSpans {
		if seen[span.TraceID] {
			continue
		}
		seen[span.TraceID] = true

		chain := traceErrorChainUpstream(stores, span)
		if len(chain) == 0 {
			continue
		}

		rootSpan := chain[len(chain)-1]
		result := ErrorChainResult{
			RootCause: &RootCauseInfo{
				Service:   rootSpan.Service,
				Operation: rootSpan.Operation,
				SpanID:    rootSpan.ID,
				TraceID:   rootSpan.TraceID,
			},
			SpanChain: chain,
			TraceID:   span.TraceID,
		}

		// Gather correlated logs
		for _, s := range chain {
			if s.IsError {
				clusters := stores.signals.LogClustersForService(s.Service)
				for _, lc := range clusters {
					if lc.LastSeen.After(since) {
						result.CorrelatedLogs = append(result.CorrelatedLogs, *lc)
					}
				}
			}
		}

		results = append(results, result)
	}

	return results
}

// traceErrorChainUpstream walks CHILD_OF edges upstream from an error span to
// the root within a single tenant's TraceStore.
func traceErrorChainUpstream(stores *tenantStores, span *SpanNode) []SpanNode {
	var chain []SpanNode
	visited := make(map[string]bool)
	current := span

	for current != nil && !visited[current.ID] {
		visited[current.ID] = true
		chain = append(chain, *current)

		if current.ParentSpanID == "" {
			break
		}
		parent, ok := stores.traces.GetSpan(current.ParentSpanID)
		if !ok {
			break
		}
		current = parent
	}

	return chain
}

// ImpactAnalysis performs BFS downstream from a service to find affected services.
func (g *GraphRAG) ImpactAnalysis(ctx context.Context, service string, maxDepth int) *ImpactResult {
	if maxDepth <= 0 {
		maxDepth = 5
	}

	stores := g.storesFor(ctx)

	result := &ImpactResult{Service: service}
	visited := map[string]bool{service: true}

	type queueItem struct {
		svc   string
		depth int
	}
	queue := []queueItem{{service, 0}}

	for len(queue) > 0 {
		item := queue[0]
		queue = queue[1:]

		if item.depth >= maxDepth {
			continue
		}

		edges := stores.service.CallEdgesFrom(item.svc)
		for _, e := range edges {
			if visited[e.ToID] {
				continue
			}
			visited[e.ToID] = true

			svc, _ := stores.service.GetService(e.ToID)
			impact := 1.0
			if svc != nil {
				impact = 1.0 - svc.HealthScore
			}

			result.AffectedServices = append(result.AffectedServices, AffectedEntry{
				Service:     e.ToID,
				Depth:       item.depth + 1,
				CallCount:   e.CallCount,
				ImpactScore: impact,
			})

			queue = append(queue, queueItem{e.ToID, item.depth + 1})
		}
	}

	result.TotalDownstream = len(result.AffectedServices)
	return result
}

// RootCauseAnalysis combines ErrorChain with anomaly correlation to rank probable causes.
func (g *GraphRAG) RootCauseAnalysis(ctx context.Context, service string, since time.Time) []RankedCause {
	stores := g.storesFor(ctx)
	errorChains := g.ErrorChain(ctx, service, since, 20)
	anomalies := stores.anomalies.AnomaliesForService(service, since)

	// Score services by how often they appear as root cause
	causeScores := make(map[string]*RankedCause)

	for _, ec := range errorChains {
		if ec.RootCause == nil {
			continue
		}
		key := ec.RootCause.Service + "|" + ec.RootCause.Operation
		rc, ok := causeScores[key]
		if !ok {
			rc = &RankedCause{
				Service:   ec.RootCause.Service,
				Operation: ec.RootCause.Operation,
			}
			causeScores[key] = rc
		}
		rc.Score += 1.0
		rc.Evidence = append(rc.Evidence, "error chain from trace "+ec.TraceID)
		if len(ec.SpanChain) > 0 {
			rc.ErrorChain = ec.SpanChain
		}
	}

	// Boost score with anomalies
	for _, a := range anomalies {
		key := a.Service + "|"
		for k, rc := range causeScores {
			if len(k) >= len(a.Service) && k[:len(a.Service)] == a.Service {
				rc.Score += 2.0
				rc.Anomalies = append(rc.Anomalies, *a)
				rc.Evidence = append(rc.Evidence, "anomaly: "+a.Evidence)
			}
		}
		// Add service itself if not seen
		if _, ok := causeScores[key]; !ok {
			causeScores[key] = &RankedCause{
				Service:   a.Service,
				Score:     2.0,
				Anomalies: []AnomalyNode{*a},
				Evidence:  []string{"anomaly: " + a.Evidence},
			}
		}
	}

	// Sort by score descending
	ranked := make([]RankedCause, 0, len(causeScores))
	for _, rc := range causeScores {
		ranked = append(ranked, *rc)
	}
	sort.Slice(ranked, func(i, j int) bool {
		return ranked[i].Score > ranked[j].Score
	})

	return ranked
}

// DependencyChain returns the full span tree for a trace.
func (g *GraphRAG) DependencyChain(ctx context.Context, traceID string) []SpanNode {
	stores := g.storesFor(ctx)
	spans := stores.traces.SpansForTrace(traceID)
	out := make([]SpanNode, len(spans))
	for i, s := range spans {
		out[i] = *s
	}
	// Sort by timestamp
	sort.Slice(out, func(i, j int) bool {
		return out[i].Timestamp.Before(out[j].Timestamp)
	})
	return out
}

// CorrelatedSignals gathers all related signals for a service within a time range.
type CorrelatedSignalsResult struct {
	Service     string             `json:"service"`
	ErrorLogs   []LogClusterNode   `json:"error_logs"`
	Metrics     []MetricNode       `json:"metrics"`
	Anomalies   []AnomalyNode      `json:"anomalies"`
	ErrorChains []ErrorChainResult `json:"error_chains,omitempty"`
}

func (g *GraphRAG) CorrelatedSignals(ctx context.Context, service string, since time.Time) *CorrelatedSignalsResult {
	stores := g.storesFor(ctx)
	result := &CorrelatedSignalsResult{Service: service}

	// Error logs
	clusters := stores.signals.LogClustersForService(service)
	for _, lc := range clusters {
		if lc.LastSeen.After(since) {
			result.ErrorLogs = append(result.ErrorLogs, *lc)
		}
	}

	// Metrics
	metrics := stores.signals.MetricsForService(service)
	for _, m := range metrics {
		result.Metrics = append(result.Metrics, *m)
	}

	// Anomalies
	anomalies := stores.anomalies.AnomaliesForService(service, since)
	for _, a := range anomalies {
		result.Anomalies = append(result.Anomalies, *a)
	}

	// Recent error chains
	result.ErrorChains = g.ErrorChain(ctx, service, since, 5)

	return result
}

// ShortestPath finds the shortest path between two services using Dijkstra.
func (g *GraphRAG) ShortestPath(ctx context.Context, from, to string) []string {
	stores := g.storesFor(ctx)
	// Build adjacency from CALLS edges
	stores.service.mu.RLock()
	adj := make(map[string]map[string]float64)
	for _, e := range stores.service.Edges {
		if e.Type != EdgeCalls {
			continue
		}
		if adj[e.FromID] == nil {
			adj[e.FromID] = make(map[string]float64)
		}
		weight := 1.0
		if e.CallCount > 0 {
			weight = 1.0 / float64(e.CallCount) // inverse call frequency
		}
		adj[e.FromID][e.ToID] = weight
		// Also add reverse for bidirectional search
		if adj[e.ToID] == nil {
			adj[e.ToID] = make(map[string]float64)
		}
		adj[e.ToID][e.FromID] = weight
	}
	stores.service.mu.RUnlock()

	// Dijkstra
	dist := map[string]float64{from: 0}
	prev := map[string]string{}
	visited := map[string]bool{}

	for {
		// Find unvisited node with minimum distance
		var u string
		minDist := math.MaxFloat64
		for node, d := range dist {
			if !visited[node] && d < minDist {
				u = node
				minDist = d
			}
		}
		if u == "" || u == to {
			break
		}
		visited[u] = true

		for neighbor, weight := range adj[u] {
			alt := dist[u] + weight
			if d, ok := dist[neighbor]; !ok || alt < d {
				dist[neighbor] = alt
				prev[neighbor] = u
			}
		}
	}

	// Reconstruct path
	if _, ok := dist[to]; !ok {
		return nil
	}
	var path []string
	for at := to; at != ""; at = prev[at] {
		path = append([]string{at}, path...)
		if at == from {
			break
		}
	}
	if len(path) > 0 && path[0] != from {
		return nil // no path
	}
	return path
}

// AnomalyTimeline returns recent anomalies sorted by time.
func (g *GraphRAG) AnomalyTimeline(ctx context.Context, since time.Time) []*AnomalyNode {
	stores := g.storesFor(ctx)
	anomalies := stores.anomalies.AnomaliesSince(since)
	sort.Slice(anomalies, func(i, j int) bool {
		return anomalies[i].Timestamp.After(anomalies[j].Timestamp)
	})
	return anomalies
}

// AnomaliesForService is a tenant-aware read-through to the per-tenant
// AnomalyStore, exported so handlers outside this package never need to
// reach into the store maps directly.
func (g *GraphRAG) AnomaliesForService(ctx context.Context, service string, since time.Time) []*AnomalyNode {
	return g.storesFor(ctx).anomalies.AnomaliesForService(service, since)
}

// AllServiceEdges returns every edge in the caller's tenant's ServiceStore.
// Kept as a narrow helper so API handlers do not need to traverse the
// tenantStores composite themselves.
func (g *GraphRAG) AllServiceEdges(ctx context.Context) []*Edge {
	return g.storesFor(ctx).service.AllEdges()
}

// ServiceMap returns the service topology with health scores for the API.
type ServiceMapEntry struct {
	Service    *ServiceNode     `json:"service"`
	Operations []*OperationNode `json:"operations,omitempty"`
	CallsTo    []*Edge          `json:"calls_to,omitempty"`
	CalledBy   []*Edge          `json:"called_by,omitempty"`
}

func (g *GraphRAG) ServiceMap(ctx context.Context, depth int) []ServiceMapEntry {
	stores := g.storesFor(ctx)
	services := stores.service.AllServices()
	result := make([]ServiceMapEntry, 0, len(services))

	for _, svc := range services {
		entry := ServiceMapEntry{
			Service:  svc,
			CallsTo:  stores.service.CallEdgesFrom(svc.Name),
			CalledBy: stores.service.CallEdgesTo(svc.Name),
		}

		// Get operations for this service
		stores.service.mu.RLock()
		for _, op := range stores.service.Operations {
			if op.Service == svc.Name {
				entry.Operations = append(entry.Operations, op)
			}
		}
		stores.service.mu.RUnlock()

		result = append(result, entry)
	}

	return result
}

File: src/graphrag/queries.py
"""GraphRAG query functions: ErrorChain, ImpactAnalysis, RootCause, Dijkstra, etc."""

from __future__ import annotations

import heapq
import math
from collections import deque
from datetime import datetime, timedelta, timezone
from typing import TYPE_CHECKING

from src.graphrag.schema import (
    AffectedEntry,
    AnomalyNode,
    CorrelatedSignalsResult,
    Edge,
    EdgeType,
    ErrorChainResult,
    ImpactResult,
    RankedCause,
    RootCauseInfo,
    ServiceMapEntry,
    SpanNode,
)

if TYPE_CHECKING:
    from src.graphrag.builder import GraphRAG


def error_chain(g: "GraphRAG", service: str, since: datetime, limit: int = 10) -> list[ErrorChainResult]:
    if limit <= 0:
        limit = 10
    error_spans = g.trace_store.error_spans(service, since)
    if len(error_spans) > limit:
        error_spans = error_spans[:limit]

    results: list[ErrorChainResult] = []
    seen: set[str] = set()

    for span in error_spans:
        if span.trace_id in seen:
            continue
        seen.add(span.trace_id)

        chain = _trace_error_chain_upstream(g, span)
        if not chain:
            continue

        root_span = chain[-1]
        result = ErrorChainResult(
            root_cause=RootCauseInfo(
                service=root_span.service,
                operation=root_span.operation,
                span_id=root_span.id,
                trace_id=root_span.trace_id,
            ),
            span_chain=chain,
            trace_id=span.trace_id,
        )

        # Correlated logs
        for s in chain:
            if s.is_error:
                clusters = g.signal_store.log_clusters_for_service(s.service)
                for lc in clusters:
                    if lc.last_seen > since:
                        result.correlated_logs.append(lc)

        results.append(result)

    return results


def _trace_error_chain_upstream(g: "GraphRAG", span: SpanNode) -> list[SpanNode]:
    chain: list[SpanNode] = []
    visited: set[str] = set()
    current: SpanNode | None = span

    while current is not None and current.id not in visited:
        visited.add(current.id)
        chain.append(current)
        if not current.parent_span_id:
            break
        current = g.trace_store.get_span(current.parent_span_id)

    return chain


def impact_analysis(g: "GraphRAG", service: str, max_depth: int = 5) -> ImpactResult:
    if max_depth <= 0:
        max_depth = 5

    result = ImpactResult(service=service)
    visited = {service}
    queue: deque[tuple[str, int]] = deque([(service, 0)])

    while queue:
        svc, depth = queue.popleft()
        if depth >= max_depth:
            continue

        edges = g.service_store.call_edges_from(svc)
        for e in edges:
            if e.to_id in visited:
                continue
            visited.add(e.to_id)

            svc_node = g.service_store.get_service(e.to_id)
            impact = 1.0
            if svc_node is not None:
                impact = 1.0 - svc_node.health_score

            result.affected_services.append(AffectedEntry(
                service=e.to_id,
                depth=depth + 1,
                call_count=e.call_count,
                impact_score=impact,
            ))
            queue.append((e.to_id, depth + 1))

    result.total_downstream = len(result.affected_services)
    return result


def root_cause_analysis(g: "GraphRAG", service: str, since: datetime) -> list[RankedCause]:
    error_chains = error_chain(g, service, since, limit=20)
    anomalies = g.anomaly_store.anomalies_for_service(service, since)

    cause_scores: dict[str, RankedCause] = {}

    for ec in error_chains:
        if ec.root_cause is None:
            continue
        key = f"{ec.root_cause.service}|{ec.root_cause.operation}"
        rc = cause_scores.get(key)
        if rc is None:
            rc = RankedCause(service=ec.root_cause.service, operation=ec.root_cause.operation)
            cause_scores[key] = rc
        rc.score += 1.0
        rc.evidence.append(f"error chain from trace {ec.trace_id}")
        if ec.span_chain:
            rc.error_chain = ec.span_chain

    for a in anomalies:
        key_prefix = f"{a.service}|"
        matched = False
        for k, rc in cause_scores.items():
            if k.startswith(a.service):
                rc.score += 2.0
                rc.anomalies.append(a)
                rc.evidence.append(f"anomaly: {a.evidence}")
                matched = True
        if not matched:
            cause_scores[key_prefix] = RankedCause(
                service=a.service,
                score=2.0,
                anomalies=[a],
                evidence=[f"anomaly: {a.evidence}"],
            )

    ranked = sorted(cause_scores.values(), key=lambda x: x.score, reverse=True)
    return ranked


def dependency_chain(g: "GraphRAG", trace_id: str) -> list[SpanNode]:
    spans = g.trace_store.spans_for_trace(trace_id)
    spans.sort(key=lambda s: s.timestamp)
    return spans


def correlated_signals(g: "GraphRAG", service: str, since: datetime) -> CorrelatedSignalsResult:
    result = CorrelatedSignalsResult(service=service)

    clusters = g.signal_store.log_clusters_for_service(service)
    for lc in clusters:
        if lc.last_seen > since:
            result.error_logs.append(lc)

    metrics = g.signal_store.metrics_for_service(service)
    for m in metrics:
        result.metrics.append(m)

    anomalies = g.anomaly_store.anomalies_for_service(service, since)
    for a in anomalies:
        result.anomalies.append(a)

    result.error_chains = error_chain(g, service, since, limit=5)
    return result


def shortest_path(g: "GraphRAG", from_svc: str, to_svc: str) -> list[str]:
    """Dijkstra shortest path between services via CALLS edges."""
    with g.service_store._lock:
        adj: dict[str, dict[str, float]] = {}
        for e in g.service_store.edges.values():
            if e.type != EdgeType.CALLS:
                continue
            weight = 1.0 / e.call_count if e.call_count > 0 else 1.0
            adj.setdefault(e.from_id, {})[e.to_id] = weight
            adj.setdefault(e.to_id, {})[e.from_id] = weight

    dist: dict[str, float] = {from_svc: 0.0}
    prev: dict[str, str] = {}
    visited: set[str] = set()

    # Simple Dijkstra with linear scan (small graphs)
    while True:
        u = ""
        min_dist = float("inf")
        for node, d in dist.items():
            if node not in visited and d < min_dist:
                u = node
                min_dist = d
        if not u or u == to_svc:
            break
        visited.add(u)
        for neighbor, weight in adj.get(u, {}).items():
            alt = dist[u] + weight
            if neighbor not in dist or alt < dist[neighbor]:
                dist[neighbor] = alt
                prev[neighbor] = u

    if to_svc not in dist:
        return []
    path: list[str] = []
    at = to_svc
    while at:
        path.insert(0, at)
        if at == from_svc:
            break
        at = prev.get(at, "")
    if not path or path[0] != from_svc:
        return []
    return path


def anomaly_timeline(g: "GraphRAG", since: datetime) -> list[AnomalyNode]:
    anomalies = g.anomaly_store.anomalies_since(since)
    anomalies.sort(key=lambda a: a.timestamp, reverse=True)
    return anomalies


def service_map(g: "GraphRAG", depth: int = 0) -> list[ServiceMapEntry]:
    services = g.service_store.all_services()
    result: list[ServiceMapEntry] = []

    for svc in services:
        entry = ServiceMapEntry(
            service=svc,
            calls_to=g.service_store.call_edges_from(svc.name),
            called_by=g.service_store.call_edges_to(svc.name),
        )
        with g.service_store._lock:
            entry.operations = [op for op in g.service_store.operations.values() if op.service == svc.name]
        result.append(entry)

    return result
